### ISSUE TITLE
Multiproof stops working starting with 9 leaves

### ISSUE THREAD

**Issue by tsudmi** [OpenZeppelin/openzeppelin-contracts#3743]
The following test reverts
```typescript
it('multiproof', async () => {
  const leaves = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i'].map(keccak256).sort(Buffer.compare)
  const merkleTree = new MerkleTree(leaves, keccak256, { sort: true })
  const root = merkleTree.getRoot()
  const proof = merkleTree.getMultiProof(leaves)
  const proofFlags = merkleTree.getProofFlags(leaves, proof)
  await merkleProof.multiProofVerifyCalldata(proof, proofFlags, root, leaves)
})
```

Versions:
- merkletreejs: `0.2.32`
- keccak256: `1.0.6`
- @openzeppelin/contracts: `4.7.3`

**Comment by frangio** [comment]
I can't reproduce this. The test passes for me. Can you share a full repo where the test fails?

**Comment by tsudmi** [comment]
Hi @frangio, thanks for the reply!

I've created a repo where the test fails for me: https://github.com/tsudmi/multiproof

Maybe some additional context about my env would be useful:
OS: Mac OS Monterey 12.4 (M1)
Node: v14.20.0

**Comment by Amxx** [comment]
I managed to replicate it. It doesn't affect all size >9.

What I've noticed:
- for some tree shapes, the test always pass.
- for some tree shapes, the result depend on the leafs selected

**Comment by Amxx** [comment]
- Given a tree shape for which there are issues, the more leaves are in the proof, the more likely the test fail.

**Comment by Amxx** [comment]
To reproduce:

```
  it('test', async function () {
    const leaves = 'abcdefghi'.split('').map(keccak256).sort(Buffer.compare);
    const merkleTree = new MerkleTree(leaves, keccak256, { sort: true });

    // This reverts with "MerkleProof: invalid multiproof"
    {
      const proofLeaves = leaves.slice(-1);
      const root = merkleTree.getHexRoot();
      const proof = merkleTree.getMultiProof(proofLeaves);
      const proofFlags = merkleTree.getProofFlags(proofLeaves, proof);
      expect(await this.merkleProof.multiProofVerifyCalldata(proof, proofFlags, root, proofLeaves)).to.be.equal(true)
    }
    // This returns false, but doesn't revert
    {
      const proofLeaves = leaves;
      const root = merkleTree.getHexRoot();
      const proof = merkleTree.getMultiProof(proofLeaves);
      const proofFlags = merkleTree.getProofFlags(proofLeaves, proof);
      expect(await this.merkleProof.multiProofVerifyCalldata(proof, proofFlags, root, proofLeaves)).to.be.equal(true)
    }
  });
```

**Comment by Amxx** [comment]
With 9 leaves, The tree looks like this:

```mermaid
graph TD;
    R-->N1; 
    N1-->N2; N1-->N3;
    N2-->N4; N2-->N5;
    N3-->N6; N3-->N7;
    N4-->L1;
    N4-->L2;
    N5-->L3;
    N5-->L4;
    N6-->L5;
    N6-->L6;
    N7-->L7;
    N7-->L8;
    R-->L9;
```

Proving L9 is part of the tree, the code expects:
- the leaf `L9`
- the internal node `N1` (that should be in the proof array)
- a flag telling it that the values should be hashed together, and not passed trough the branch merging algorithm. (in the proofFlags array)

However, when running `merkleTree.getProofFlags(proofLeaves, proof)`, the array is empty. This causes the sanity check 
```
require(leavesLen + proof.length - 1 == totalHashes, "MerkleProof: invalid multiproof");
```
to fail.

While this is not exactly @tsudmi 's issue, I believe they are related.


That is using `merkletreejs@0.2.32`

**Comment by Amxx** [comment]
Upon further investigation, it turns out @tsudmi's issue is mostly unrelated.

--- 

TLDR: the shape of the trees produced by the library are (sometimes) not compatible with the verification algorithm.

---

Taking our example above, if the leaves are selected, we get no internal node as part of the proof. What the algorithm then does is
```
hash(L1,L2) → temp1 (N4)
hash(L3,L4) → temp2 (N5)
hash(L5,L6) → temp3 (N6)
hash(L7,L8) → temp4 (N7)
... so far so good ...
hash(L9,temp1) → temp5
... this is bad
```
this is because of the rule that is "unless the flag tells us to read from the proof nodes, we process leaves then internally reconstructed nodes, in order"

AFAIK, it should possible to organize the tree so that it is possible to produce an arbitrary multiproof but it is not that simple, and the library doesn't do it.

**Comment by tsudmi** [comment]
Hi @Amxx, thanks for diving into this. Seems like it really depends on the tree shape here, as I was also unable to verify with the last element and specific subsets. Does it mean that the fix should be in the `merkletreejs` library, or is it the verifying contract that should be modified?

**Comment by Amxx** [comment]
Note: one of the assumption the current multiproof verification algorithm makes, which is not verified is that:

- For a tree of depth N
- Leaves are either at depth N or N-1.
- Leaves that are at depth N are processed first (on the left), and leaves that are at depth N-1 are processed after (on the right).

The tree of width 9 does not verify this assumption.

This is a weak version of the assumption, as shown above there are more assumption about the order in which leaves are processed during reconstruction.

**Comment by frangio** [comment]
We have made progress with this issue and have a good idea for how to move forward. Will share more details soon.

**Comment by tsudmi** [comment]
Thanks, @frangio @Amxx , looking forward to the fix. Let me know if there is anything I can help with.

**Comment by frangio** [comment]
The algorithm for multiproofs doesn't allow the construction of multiproofs for trees with particularly unbalanced shapes. The merkletree.js library can currently easily produce such trees, as in the example above. Unfortunately this is not something we knew of before.

The problem is solved by better balancing the merkle tree. The tree should be a complete (not perfect) full binary tree (definitions [in Wikipedia](https://en.wikipedia.org/wiki/Binary_tree)). This kind of tree exists for any needed total number of leaves.

Additionally, the array of leaves that will be proven with a multiproof needs to be sorted in a particular way to be able to produce a multiproof. The ordering should be the reverse of the "general indices" defined in [this document](https://github.com/ethereum/consensus-specs/blob/6b9e65fab6c2cf1f8973b3dee29173ed6cb6130b/ssz/merkle-proofs.md#generalized-merkle-tree-index). The proof array will be sorted in the same way.

The plan I propose is:
1. Document that in order to use multiproofs there are requirements on the tree shape and the leaves array. (Note: This kind of requirement is not entirely new, because even for single proofs you want the tree to be well balanced as well.)
2. Add a flag in merkletree.js to build trees of this shape, add a warning if a multiproof is requested for a tree that was built without this flag, and verify that leaves are properly ordered when a multiproof is requested.

**Comment by tsudmi** [comment]
@frangio Would it make sense to add a test case for that?

**Comment by frangio** [comment]
Yes we will want to add a test case. At the moment the next step we're working on is to modify the merkletreejs library to construct the tree as needed.

**Comment by tsudmi** [comment]
Hi @frangio, I've tested the `complete` parameter added in https://github.com/merkletreejs/merkletreejs/pull/68. It looks like the complete tree does not pass the proof validity here: https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/utils/cryptography/MerkleProof.sol#L183

**Comment by Amxx** [comment]
Yes @tsudmi that is something we are aware of. Just having a complete tree is not enough, there additional constrain that consider when constructing the tree.

**Comment by frangio** [comment]
The merkletreejs library has a large scope and we were not able or comfortable enough with all aspects of it to fix multiproof generation in all situations... We understand that the `complete` parameter is not enough to make it generate valid multiproofs, but we know that it is at the very least required to give the tree that shape.

To fully fix this issue with multiproofs, we've decided to release our own JavaScript merkle tree library, with a smaller scope. It will be our recommended companion JavaScript library moving forward and essentially a successor to merkletreejs. It's already being developed and we want to release it as soon as possible but we need to make sure it is well tested first. @tsudmi If you're interested we can give you early access.

**Comment by tsudmi** [comment]
Hi, @frangio it would be great to get early access. I have some tests that I could run with it.

**Comment by frangio** [comment]
The library is released now. Multiproofs generated by this library can be successfully validated by the Solidity code.

https://github.com/OpenZeppelin/merkle-tree
